dawn/prototype/AtlasFromNetcdf.cpp

```diff
@@ -0,0 +1,298 @@
+  atlas::array::ArrayView<atlas::gidx_t, 1> glb_idx_cell =
+      atlas::array::make_view<atlas::gidx_t, 1>(mesh.cells().global_index());
+
+  for(int cellIdx = 0; cellIdx < ncells; cellIdx++) {
+    // indices in netcdf are 1 based, data is column major
+    atlas::idx_t tri_nodes[3] = {cellToVertex[0 * ncells + cellIdx] - 1,
+                                 cellToVertex[1 * ncells + cellIdx] - 1,
+                                 cellToVertex[2 * ncells + cellIdx] - 1};
+    node_connectivity.set(cellIdx, tri_nodes);
+    glb_idx_cell[cellIdx] = cellIdx;
+    cells_part(cellIdx) = myPart;
+  }
+
+  return true;
+}
+
+static bool AddNeighborList(const netCDF::NcFile& dataFile, atlas::Mesh& mesh,
+                            const std::string nbhListName, int yPerXExpected,
+                            atlas::mesh::HybridElements::Connectivity& connectivity) {
+  auto [xToY, yPerX, numY] = Load2DField<int>(dataFile, nbhListName);
+  if(yPerX != yPerXExpected) {
+    std::cout << "number of neighbors per element not as expected!\n";
+    return false;
+  }
+  for(int elemIdx = 0; elemIdx < numY; elemIdx++) {
+    atlas::idx_t yOfX[yPerX];
+    for(int innerIdx = 0; innerIdx < yPerX; innerIdx++) {
+      // indices in netcdf are 1 based, data is column major
+      yOfX[innerIdx] = xToY[innerIdx * numY + elemIdx] - 1;
+    }
+    connectivity.set(elemIdx, yOfX);
+  }
+  return true;
+}
+
+static bool AddNeighborList(const netCDF::NcFile& dataFile, atlas::Mesh& mesh,
+                            const std::string nbhListName, int yPerXExpected,
+                            atlas::mesh::Nodes::Connectivity& connectivity) {
+  auto [xToY, yPerX, numY] = Load2DField<int>(dataFile, nbhListName);
+  if(yPerX != yPerXExpected) {
+    std::cout << "number of neighbors per element not as expected!\n";
+    return false;
+  }
+  for(int elemIdx = 0; elemIdx < numY; elemIdx++) {
+    atlas::idx_t yOfX[yPerX];
+    for(int innerIdx = 0; innerIdx < yPerX; innerIdx++) {
+      // indices in netcdf are 1 based, data is column major
+      yOfX[innerIdx] = xToY[innerIdx * numY + elemIdx] - 1;
+    }
+    connectivity.set(elemIdx, yOfX);
+  }
+  return true;
+}
+
+// Pre-allocate neighborhood table, snippet taken from
+// https://github.com/ecmwf/atlas/blob/98ff1f20dc883ba2dfd7658196622b9bc6d6fceb/src/atlas/mesh/actions/BuildEdges.cc#L73
+static void AllocNbhTable(atlas::mesh::HybridElements::Connectivity& connectivity, int numElements,
+                          int nbhPerElem) {
+  std::vector<int> init(numElements * nbhPerElem, connectivity.missing_value());
+  connectivity.add(numElements, nbhPerElem, init.data());
+}
+
+static void AllocNbhTable(atlas::mesh::Nodes::Connectivity& connectivity, int numElements,
+                          int nbhPerElem) {
+  std::vector<int> init(numElements * nbhPerElem, connectivity.missing_value());
+  connectivity.add(numElements, nbhPerElem, init.data());
```

Comment:
This should create a dangling ptr to `init.data()`. Or do I miss something?

Reply:
The naming in atlas is confusing here. `Add` copies the data in init: https://github.com/ecmwf/atlas/blob/develop/src/atlas/mesh/Connectivity.cc#L139

Reply:
No, need to change, but it seems that just calling resize would also fill with `missing_value()`?

Reply:
yes, however, the `add` functions also resizes some internal structures and fields. I'll leave as is because I'm afraid to break something down the line.